docs/spec/fastsync/scheduler.tla

```diff
@@ -0,0 +1,604 @@
+------------------------------- MODULE scheduler -------------------------------
+(*
+ A specification of the fast sync scheduler that is introduced in blockchain/v2:
+ 
+ https://github.com/tendermint/tendermint/tree/brapse/blockchain-v2-riri-reactor-2
+ 
+ The model includes:
+    - a scheduler that maintains the peers and blocks that it receives from the peers, and
+    - one environment simulating a correct peer
+    
+ This specification focuses on the events that are received and produced by the scheduler.
+ Communication between the scheduler and the other fastsync components is not specified.
+*)
+
+EXTENDS Integers, FiniteSets
+
+\* the protocol parameters
+CONSTANTS
+    PeerIDs,        \* potential peer ids, a set of integers, e.g., 0..2
+    ultimateHeight, \* the maximum height of the blockchain, an integer, e.g., 3
+    numRequests     \* the number of requests made per batch, e.g., 2
+
+\* a few definitions
+None == -1                      \* an undefined value
+Heights == 0..ultimateHeight    \* potential heights
+noErr == "errNone"
+Errors == {
+  noErr, "errPeerNotFound", "errDelRemovedPeer", "errAddDuplicatePeer",
+  "errUpdateRemovedPeer", "errAfterPeerRemove", "errBadPeer", "errBadPeerState",
+  "errProcessedBlockEv", "finished", "timeout", "errAddRemovedPeer", "errPeerNoBlock", "errBadBlockState"}
+
+PeerStates == {"peerStateUnknown", "peerStateNew", "peerStateReady", "peerStateRemoved"}
+
+BlockStates == {
+  "blockStateUnknown", "blockStateNew", "blockStatePending",
+  "blockStateReceived", "blockStateProcessed"}
+
+\* basic stuff
+Min(a, b) == IF a < b THEN a ELSE b
+Max(a, b) == IF a > b THEN a ELSE b
+
+\* the state of the scheduler:
+VARIABLE  turn          \* who makes a step: the scheduler or the environment
+
+\* the state of the reactor: 
+VARIABLES inEvent,       \* an event from the environment to the scheduler
+          envRunning \* a Boolean, negation of stopProcessing in the implementation
+                    
+\* the state of the scheduler:
+VARIABLES outEvent,       \* an event from the scheduler to the environment
+          scRunning
+
+\* the block pool: 
+VARIABLE scheduler
+    (*
+       scheduler is a record that contains: 
+         height: Int,
+            height of the next block to collect
+         peers: PeerIDs,
+            the set of peers that have connected in the past, may include removed peers
+         peerHeights: [PeerIDs -> Heights],
+            a map to collect the peer heights, >0 if peer in ready state, None (-1) otherwise
+         peerStates: [PeerIDs -> PeerStates],
+            a map to record the peer states 
+         blockStates: [Heights -> BlockStates]
```

Comment:
There is also a field sc.blocks that is used. You might want to add it here with a description.

Reply:
added